juju/framework.py

```diff
@@ -702,6 +702,46 @@ def insert(self, index, value):
     def append(self, value):
         self._under.append(value)
 
+    def __eq__(self, other):
+        if isinstance(other, StoredList):
+            return self._under == list(other)
```

Comment:
This allocates and copies a reference to every element of other just to test for equality. That means the cost is not proportional to the code. I suggest leaving that issue aside for now.

Reply:
I could instead use `other._under`. While this is not a good idea when working with a different type, we are implementing a comparison inside a `StoredList`'s `__eq__` with another `StoredList` - it seems normal to rely on the implementation details of `other` here. So long as we can guarantee that `list(other) == other._under` is always True in any subclass (not sure if we will ever have any).

Reply:
Yeah, that sounds okay. If we go that way, then this PR just needs to address testing and also change the `return NotImplemented` to `return False`, which is the usual behavior I believe.

Reply:
I used `NotImplemented` following the docs https://docs.python.org/3/library/constants.html#NotImplemented to allow for the standard fallback behavior:

> When a binary (or in-place) method returns NotImplemented the interpreter will try the reflected operation on the other type (or some other fallback, depending on the operator). If all attempts return NotImplemented, the interpreter will raise an appropriate exception. Incorrectly returning NotImplemented will result in a misleading error message or the NotImplemented value being returned to Python code.

Do you think it does not make sense in this case?

Reply:
Sounds good then. It still works as one would expect:
```
>>> class C:                                                                                                                                                             
...     def __eq__(self, other):                                                                                                                                         
...         return NotImplemented                                                                                                                                        
...                                                                                                                                                                      
>>> o = C()                                                                                                                                                              
>>> o == []                                                                                                                                                              
False                                                                                                                                                                    
>>> [] == o                                                                                                                                                              
False
```